src/java/org/apache/cassandra/cql3/restrictions/ClusteringElements.java

```diff
@@ -0,0 +1,473 @@
+        if (!(columns.get(0) instanceof ColumnMetadata))
+            throw new UnsupportedOperationException("Non partition key or clustering columns cannot be extended");
+
+        if (!suffix.isEmpty()) // suffix can be empty if equal to (top) or (bottom)
+        {
+            ColumnMetadata lastPrefixElement = ((ColumnMetadata) last(this.columns));
+            ColumnMetadata firstSuffixElement = ((ColumnMetadata) suffix.columns.get(0));
+            if (firstSuffixElement.kind != lastPrefixElement.kind)
+                throw new UnsupportedOperationException("Cannot extend elements with elements of a different kind");
+            if (firstSuffixElement.position() != lastPrefixElement.position() + 1)
+                throw new UnsupportedOperationException("Cannot extend elements with non consecutive elements");
+        }
+    }
+
+    private static <T> ImmutableList<T> concat(ImmutableList<? extends T> prefix, ImmutableList<? extends T> suffix)
+    {
+        return ImmutableList.<T>builderWithExpectedSize(prefix.size() + suffix.size())
+                            .addAll(prefix)
+                            .addAll(suffix)
+                            .build();
+    }
+
+    /**
+     * Returns a {@code RangeSet} representing all the possible {@code ClusteringElements} values.
+     * @return a {@code RangeSet} representing all the possible {@code ClusteringElements} values.
+     */
+    public static RangeSet<ClusteringElements> all()
+    {
+        TreeRangeSet<ClusteringElements> rangeSet = TreeRangeSet.create();
+        rangeSet.add(ALL);
+        return rangeSet;
+    }
+
+    /**
+     * Returns a {@code RangeSet} that contains all values less than or equal to endpoint.
+     * @return a {@code RangeSet} that contains all values less than or equal to endpoint.
+     */
+    public static RangeSet<ClusteringElements> atMost(ClusteringElements endpoint)
+    {
+        return buildRangeSet(endpoint, true, BoundType.CLOSED);
+    }
+
+    /**
+     * Returns a {@code RangeSet} that contains all values less than endpoint.
+     * @return a {@code RangeSet} that contains all values less than endpoint.
+     */
+    public static RangeSet<ClusteringElements> lessThan(ClusteringElements endpoint)
+    {
+        return buildRangeSet(endpoint, true, BoundType.OPEN);
+    }
+
+    /**
+     * Returns a {@code RangeSet} that contains all values greater or equal to endpoint.
+     * @return a {@code RangeSet} that contains all values greater or equal to endpoint.
+     */
+    public static RangeSet<ClusteringElements> atLeast(ClusteringElements endpoint)
+    {
+        return buildRangeSet(endpoint, false, BoundType.CLOSED);
+    }
+
+    /**
+     * Returns a {@code RangeSet} that contains all values greater than endpoint.
+     * @return a {@code RangeSet} that contains all values greater than endpoint.
```

Comment:
super nit, but I'd be fine w/ just leaving the `@return` line here, as the first line is redundant?

Reply:
It is just in case some day someone generate the javadoc. I worked for a company where you had to feel each part. It stuck with me.